metadata-ingestion/src/datahub/ingestion/source/aerospike.py

```diff
@@ -0,0 +1,591 @@
+            the Aerospike client
+        as_set:
+            the Aerospike set
+        delimiter:
+            string to concatenate field names by
+        records_per_second:
+            number of records per second for Aerospike query
+        sample_size:
+            number of items in the set to sample
+            (reads entire set if not provided)
+    """
+
+    query = client.query(as_set.ns, as_set.set)
+    if sample_size:
+        query.max_records = sample_size
+    query.records_per_second = records_per_second
+
+    res = query.results()
+    records = [{**record[2], "PK": record[0][2]} for record in res]
+    return construct_schema(records, delimiter)
+
+
+@platform_name("Aerospike")
+@config_class(AerospikeConfig)
+@support_status(SupportStatus.TESTING)
+@capability(SourceCapability.PLATFORM_INSTANCE, "Enabled by default")
+@capability(SourceCapability.SCHEMA_METADATA, "Enabled by default")
+@dataclass
+class AerospikeSource(StatefulIngestionSourceBase):
+    """
+    This plugin extracts the following:
+
+    - Namespaces and associated metadata
+    - Sets in each namespace and schemas for each set (via schema inference)
+
+    By default, schema inference samples 1,000 documents from each set. Setting `schemaSamplingSize: null` will scan the entire set.
+
+    Note that `schemaSamplingSize` has no effect if `enableSchemaInference: False` is set.
+
+    Really large schemas will be further truncated to a maximum of 300 schema fields. This is configurable using the `maxSchemaSize` parameter.
+
+    """
+
+    config: AerospikeConfig
+    report: AerospikeSourceReport
+    aerospike_client: aerospike.Client
+    platform: str = "aerospike"
+
+    def __init__(self, ctx: PipelineContext, config: AerospikeConfig):
+        super().__init__(config, ctx)
+        self.config = config
+        self.report = AerospikeSourceReport()
+
+        hosts = [
+            (x[0], int(x[1]), x[2] if len(x) > 2 else None) for x in self.config.hosts
+        ]
+        client_config: Dict[str, Any] = {"hosts": hosts}
+        if self.config.username is not None:
+            client_config["user"] = self.config.username
+        if self.config.password is not None:
+            client_config["password"] = self.config.password
+        if self.config.auth_mode is not None:
+            client_config["auth_mode"] = self.config.auth_mode
+        if self.config.login_timeout_ms is not None:
+            client_config["login_timeout_ms"] = self.config.login_timeout_ms
+        if self.config.tls_enabled is not None:
+            client_config["tls"] = {}
+            client_config["tls"]["enable"] = self.config.tls_enabled
+            if self.config.tls_capath is not None:
+                client_config["tls"]["capath"] = self.config.tls_capath
+            if self.config.tls_cafile is not None:
+                client_config["tls"]["cafile"] = self.config.tls_cafile
+
+        self.aerospike_client = aerospike.client(client_config).connect()
```

Comment:
needs try-except handling with self.report.report_failure if you are unable to connect (with clear error "message" and "title" fields which will inform the user of what has gone wrong -- e.g. bad username, bad password, etc)